### PR TITLE
Create Region and Site Location Types and Locations during data migration

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes: #2955 
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/dcim/templates/dcim/inc/location_hierarchy.html [hunk 1]
```diff
@@ -33,10 +23,6 @@
         {% endfor %}
                 </ul>
             </li>
```
Comment (glennmatthews) [inline nautobot/dcim/templates/dcim/inc/location_hierarchy.html:25]: Possibly need to remove these two lines as well?

File: nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                parent=Location.objects.get(name=region.parent.name, location_type=region_lt)
+                if region.parent
+                else None,
+            )
+        site_lt = LocationType.objects.create(name="Site", parent=LocationType.objects.get(name="Region"))
+        Location.objects.create(
+            location_type=region_lt,
+            name="Global Region",
+            description=region.description,
+        )
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py:30]: Maybe only create the Global Region if `Site.objects.filter(region__isnull=True).exists()`?

File: nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                contact_email=site.contact_email,
+                comments=site.comments,
+                status=site.status,
+                tags=site.tags,
+            )
+        for site in Site.objects.filter(region__isnull=False):
+            Location.objects.create(
+                name=site.name,
+                location_type=site_lt,
+                parent=Location.objects.get(location_type=region_lt, name=site.region.name),
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py:56]: Possibly merge these two for loops since this is the only line that differs, something like:

```suggestion
                parent=Location.objects.get(location_type=region_lt, name=site.region.name if site.region else "Global Region"),
```

File: nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        for location_type in LocationType.objects.filter(parent__isnull=True).exclude(name="Region"):
+            location_type.parent = site_lt
+            location_type.save()
+        for location in Location.objects.filter(site__isnull=False):
+            location.parent = Location.objects.get(name=location.site.name, location_type=site_lt)
+            location.save()
+    # Only Region instances exist
+    elif Region.objects.exists():
+        LocationType.objects.create(name="Region", nestable=True)
+        region_lt = LocationType.objects.create(name="Region", nestable=True)
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py:82]: duplicate line?

**Review by glennmatthews — COMMENTED** [review]
Looks like a promising start!

File: nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        for vlangroup in VLANGroup.objects.filter(location__isnull=True, site__isnull=False):
+            vlangroup.location = Location.objects.get(name=vlangroup.site.name, location_type=site_lt)
+            vlangroup.save()
+        for vlan in VLAN.objects.filter(location__isnull=True, site__isnull=False):
+            vlan.location = Location.objects.get(name=vlan.site.name, location_type=site_lt)
+            vlan.save()
+        for cluster in Cluster.objects.filter(location__isnull=True, site__isnull=False):
+            cluster.location = Location.objects.get(name=cluster.site.name, location_type=site_lt)
+            cluster.save()
+
```
Comment (timizuoebideri1) [inline nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py:137]: Refactor suggestion, My function names might be bad 👀:
```python

def create_region_location(region_class, location_class, region_lt):
     """
    Creates location objects for each region instance in the region_class model.

    Args:
        region_class: The model class for the regions.
        location_class: The model class for the locations.
        region_location_type: The location type for the regions.
    """
    # Breadth First Query
    regions = region_class.objects.with_tree_fields().extra(order_by=["__tree.tree_depth", "__tree.tree_ordering"])
    locations_instances = []
    for region in regions:
        parent = (
            location_class.objects.get(name=region.parent.name, location_type=region_lt)
            if region.parent else None
            )
        locations_instances.append(
            location_class(
                location_type=region_lt,
                name=region.name,
                description=region.description,
                parent=parent
            )
        )
    location_class.objects.bulk_create(locations_instances, batch_size=1000)
    

def create_site_location(
    site_class, 
    location_type_class, 
    location_class, 
    site_lt, 
    exclude__name, 
    has_parent_location=False, 
    region_lt=None,
    ):
    """
    Creates location objects for each site instance in the site_class model.

    Args:
        site_class: The model class for the sites.
        location_type_class: The model class for the location types.
        location_class: The model class for the locations.
        site_location_type: The location type for the sites.
        exclude_name: The name to exclude from the update query.
        has_parent_location: Whether the sites have a parent location or not.
        region_location_type: The location type for the regions, required if has_parent_location is True.
    """
    
    locations_instances = []
    for site in site_class.objects.all():
        extra_kwargs = {}
        if has_parent_location:
            extra_kwargs["parent"] = location_class.objects.get(
                location_type=region_lt,
                name=site.region.name if site.region else "Global Region"
            )
        locations_instances.append(
            location_class(
                name=site.name,
                location_type=site_lt,
                tenant=site.tenant,
                facility=site.facility,
                asn=site.asn,
                time_zone=site.time_zone,
                description=site.description,
                physical_address=site.physical_address,
                shipping_address=site.shipping_address,
                latitude=site.latitude,
                longitude=site.longitude,
                contact_name=site.contact_name,
                contact_phone=site.contact_phone,
                contact_email=site.contact_email,
                comments=site.comments,
                status=site.status,
                tags=site.tags,
            )
        )
    location_class.objects.bulk_create(locations_instances, batch_size=1000)
    location_class.objects.filter(site__isnull=False).update(parent=F('site__name'), location_type=site_lt)
    location_type_class.objects.filter(parent__isnull=True).exclude(name=exclude__name).update(parent=site_lt)
    

def create_region_and_site_location_types(apps, schema_editor):
    Region = apps.get_model("dcim", "region")
    Site = apps.get_model("dcim", "site")
    LocationType = apps.get_model("dcim", "locationtype")
    Location = apps.get_model("dcim", "location")

    # Region instances exist
    if Region.objects.exists():
        region_lt = LocationType.objects.create(name="Region", nestable=True)
        create_region_location(
            region_class=Region, 
            location_type_class=LocationType, 
            location_class=Location,
            region_lt=region_lt,
        )
        if Site.objects.exists():  # Both Site and Region instances exist
            parent_location = LocationType.objects.get(name="Region")
            site_lt = LocationType.objects.create(name="Site", parent=parent_location)
            if Site.objects.filter(region__isnull=True).exists():
                Location.objects.create(
                    location_type=region_lt,
                    name="Global Region",
                    description="Parent Location of Region LocationType for all sites that "
                    "did not have a region attribute set before the migration",
                )
            create_site_location(
                location_class=Location,
                location_type_class=LocationType,
                site_class=Site,
                site_lt=site_lt,
                exclude__name="Region",
                has_parent_location=True,
                region_lt=region_lt,
            )


    elif Site.objects.exists():  # Only Site instances exist, we make Site the top level LocationType
        site_lt = LocationType.objects.create(name="Site")
        create_site_location(
            site_class=Region, 
            location_type_class=LocationType, 
            location_class=Location,
            site_lt=site_lt,
            exclude__name="Site",
        )

```

Comment (jathanism) [reply nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py:137]: I second this. My feedback was going to be that the existing function is hard to follow because it's so nested and also results in undesirable duplication of code.

**Review by jathanism — CHANGES_REQUESTED** [review]
Please consider the refactor of the data migration function as suggested by @timizuoebideri1

File: nautobot/docs/release-notes/version-2.0.md [hunk 6]
```diff
@@ -18,6 +18,15 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+
```
Comment (gsnider2195) [inline nautobot/docs/release-notes/version-2.0.md:29]: This is a lot of information. Maybe we should include examples for 3 and 4?

File: nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py (hunk 5 above)
Comment (HanlinMiao) [reply nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py:137]: Just some modifications I have to make  with the code:
```
for region in regions:
        parent = (
            location_class.objects.get(name=region.parent.name, location_type=region_lt)
            if region.parent else None
          )
```
`location_class.objects.get(name=region.parent.name, location_type=region_lt)` will always throw a ` no matching query exist` error because the Region Location Type locations are created until later. 
So I changed it back to creating one Region Location at a time.
```
regions = region_class.objects.with_tree_fields().extra(order_by=["__tree.tree_depth", "__tree.tree_ordering"])
for region in regions:
        location_class.objects.create(
            location_type=region_lt,
            name=region.name,
            description=region.description,
            parent=location_class.objects.get(name=region.parent.name, location_type=region_lt)
            if region.parent
            else None,
       )
```

`location_class.objects.filter(site__isnull=False).update(parent=F('site__name'), location_type=site_lt)` This line will throw an error below:
Screen Shot 2023-01-04 at 2 46 57 PM
So I have to change it back to
```
for location in location_class.objects.filter(site__isnull=False):
        location.parent = location_class.objects.get(location_type=site_lt, name=location.site.name)
        location.save()
```

File: nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                status=site.status,
+                tags=site.tags,
+                **extra_kwargs,
+            )
+        )
+    location_class.objects.bulk_create(location_instances, batch_size=1000)
+    # Set existing top level locations to have site locations as their parents
+    for location in location_class.objects.filter(site__isnull=False):
+        location.parent = location_class.objects.get(name=location.site.name, location_type=site_lt)
+        location.save()
```
Comment (HanlinMiao) [inline nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py:84]: Wonder if there is a better way to do this with less performance penalty.

File: nautobot/docs/release-notes/version-2.0.md [hunk 8]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
 
 #### Collapse Region and Site Models into Location ([#2517](https://github.com/nautobot/nautobot/issues/2517))
 
+##### Initial Data Migration
+
+`Site` and `Region` models are removed in v2.0. As a result, the existing `Site` and `Region` data will be moved to `LocationTypes` and `Locations`. Here is what to expect:
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:23]:
```suggestion
The `Site` and `Region` models have been removed in v2.0 and have been replaced with `Location` of specific `LocationType`. As a result, the existing `Site` and `Region` data will be migrated to corresponding `LocationType` and `Location` objects. Here is what to expect:
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 9]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+
+1. If you do not have any `Site` and `Region` instances in your existing database, running this data migration will do nothing.
+2. If you only have `Region` instances in your existing database, a `Region` LocationType will be created and for each legacy `Region` instance, a corresponding `Location` instance with the same attributes (`name`, `description` and etc) will be created.
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:26]:
```suggestion
2. If you only have `Region` instances in your existing database, a `LocationType` named "Region" will be created and for each legacy `Region` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, etc.) will be created.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 10]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+3. If you only have `Site` instances in your existing database, a `Site` LocationType will be created and every root level `LocationType` in your database will have the new `Site` LocationType as their parent. For each legacy `Site` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude` and etc) will be created. In addition to that, models (`CircuitTermination`, `Device`, `PowerPanel`, `RackGroup`, `Rack`, `Prefix`, `VLANGroup`, `VLAN`, `Cluster`) with an old `Site` instance assigned without its `location` attribute specified will have its `location` point to the new `Location` of `Site` LocationType. These Models with `location` attribute specified will remain unchanged.
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:27]:
```suggestion
3. If you only have `Site` instances in your existing database, a `LocationType` named "Site" will be created and every root level `LocationType` in your database will have the new "Site" `LocationType` as their parent. For each legacy `Site` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created. In addition to that, models (`CircuitTermination`, `Device`, `PowerPanel`, `RackGroup`, `Rack`, `Prefix`, `VLANGroup`, `VLAN`, `Cluster`) with an old `Site` instance assigned without its `location` attribute specified will have its `location` point to the new `Location` of "Site" `LocationType`. All other attributes on these models will remain unchanged.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 11]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+
+    For Example:
+        A `Site` instance with name **AMS01** is the base site for two top level locations with names **root-01** and **root-02** respectively in the your existing database. During the data migration, `Site` LocationType and a location of this LocationType named **AMS01** with all the information (`asn`, `latitude`, etc) will be created and **root-01** and **root-02** will have this location as their `parent`.
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:30]:
```suggestion
        We will start with a `Site` instance with name **AMS01** as the base `Site` for two top-level `Location` objects with names **root-01** and **root-02** respectively. During the data migration, a `LocationType` named **Site** will be created, and a `Location` of **Site** `LocationType` named **AMS01** with all the information (`asn`, `latitude`, etc.) from the base `Site` will be created. The `Location` objects named **root-01** and **root-02** will have this **AMS01** `Location` set as their `parent`.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 12]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+        In addition, legacy `Site` instance with name **AMS01** also has three `Device` instances associated with it: **ams01-edge-01**, **ams01-edge-02**, **ams01-edge-03**. However, **ams01-edge-01** only has its `site` attribute set as **AMS01** whereas **ams01-edge-02** and **ams01-edge-03** have both its `site` and `location` attributes set as **AMS01** and **root-01** respectively. During the data migration, **ams01-edge-01**'s `location` attribute will point to the new location of `Site` LocationType with name **AMS01** while devices **ams01-edge-02** and **ams01-edge-03** remain unchanged.
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:31]:
```suggestion
        In addition, the legacy `Site` instance with name **AMS01** also has three `Device` instances associated with it: **ams01-edge-01**, **ams01-edge-02**, **ams01-edge-03**. However, **ams01-edge-01** only has its `site` attribute set as **AMS01** whereas **ams01-edge-02** and **ams01-edge-03** have both its `site` and `location` attributes set as **AMS01** and **root-01** respectively. During the data migration, `location` attribute for the `Device` named **ams01-edge-01**'s will point to the new `Location` named **AMS01** of `LocationType` **Site**, while devices **ams01-edge-02** and **ams01-edge-03** will remain unchanged.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 13]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
...y `Region` instance, a corresponding `Location` instance with the same attributes (`name`, `description` and etc) will be created. A `Site` LocationType will be created with the new `Region` LocationType as the parent and every root level `LocationType` in your database will have the new `Site` LocationType as a parent. For each legacy `Site` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude` and etc) will be created. If you have `Site` instances in your database without a `Region` assigned to them, one additional location named "Global Region" of LocationType `Region` will be created and all the locations of LocationType `Site` created from the legacy region-less `Site` instances will have "Global Region" as their parent. In addition to that, models (`CircuitTermination`, `Device`, `PowerPanel`, `RackGroup`, `Rack`, `Prefix`, `VLANGroup`, `VLAN`, `Cluster`) with an old `Site` instance assigned but without its `location` attribute specified will have its `location` point to the new `Location` of `Site` LocationType. These Models with `location` attribute specified will remain unchanged.
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:33]:
```suggestion
4. If you have both `Site` and `Region` instances in your existing database, a `LocationType` named **Region** will be created and for each legacy `Region` instance. A corresponding `Location` instance with the same attributes (`name`, `description`, etc.) for each `Region` will be created. A `LocationType` named **Site** will be created with the new `LocationType` named **Region** set as its `parent` and every root-level `LocationType` in your database will have the new `LocationType` named **Site** as its `parent`. For each legacy `Site` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created. If you have `Site` instances in your database without a `Region` assigned to them, one additional `Location` named **Global Region** of `LocationType` **Region** will be created and each `Location` of `LocationType` **Site** created from the legacy region-less `Site` instances will have the **Global Region** `Location` as their parent. In addition to that, models (`CircuitTermination`, `Device`, `PowerPanel`, `RackGroup`, `Rack`, `Prefix`, `VLANGroup`, `VLAN`, `Cluster`) with an old `Site` instance assigned but without its `location` attribute specified will have its `location` point to the new `Location` of `LocationType` **Site**. All other attributes on these models will remain unchanged.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 14]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+
+    For Example:
+        There are two `Site` instances and one `Region` instance in your existing database:
+        The `Region` with name **America** has one child `Site` instance named **AMS01**. And the other `Site` instance named **AUS01** is not associated with any regions (`region` attribute is set to `None`). Site **AMS01** is the base site for two top level locations with names **root-01** and **root-02** respectively in the your existing database. During the data migration, `Region` LocationType and a location of this LocationType named **America** with all the information will be created. `Site` LocationType with its `parent` set as the new LocationType `Region` and a location of `Site` LocationType named **AMS01** with all the information (`asn`, `latitude`, etc) will be created. Location **AMS01** will have Location **America** as its `parent` and locations **root-01** and **root-02** will have location **AMS01** as their `parent`. As for `Site` instance **AUS01**, since it does not have a `Region` instance associated with it, its corresponding location **AUS01** will have a new location named **Global Region** of LocationType `Region` as its parent.
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:38]:
```suggestion
        There are two `Site` instances and one `Region` instance in your existing database. The `Region` with name **America** has one child `Site` instance named **AMS01**. And the other `Site` instance named **AUS01** is not associated with any `Region` (`region` attribute is set to `None`). The `Site` **AMS01** is the base `Site` for two top-level `Location` objects with names **root-01** and **root-02** respectively. During the data migration, a `LocationType` named **Region** and a `Location` of this `LocationType` named **America** with all the same information will be created.  The `LocationType` named **Site** with its `parent` set as the new `LocationType` **Region** and a `Location` of `LocationType` named **AMS01** with all the same information (`asn`, `latitude`, etc.) will be created. The `Location` **AMS01** will have `Location` **America** as its `parent` and each `Location` **root-01** and **root-02** will have `Location` **AMS01** as its `parent`. Finally, the `Site` instance **AUS01**, since it does not have a `Region` instance associated with it, its corresponding `Location` **AUS01** will have a new `Location` named **Global Region** of `LocationType` **Region** as its `parent`.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 15]
```diff
@@ -18,6 +18,26 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+        In addition, legacy `Site` instance with name **AMS01** also has three `Device` instances associated with it: **ams01-edge-01**, **ams01-edge-02**, **ams01-edge-03**. However, **ams01-edge-01** only has its `site` attribute set as **AMS01** whereas **ams01-edge-02** and **ams01-edge-03** have both its `site` and `location` attributes set as **AMS01** and **root-01** respectively. During the data migration, **ams01-edge-01**'s `location` attribute will point to the new location of `Site` LocationType with name **AMS01** while devices **ams01-edge-02** and **ams01-edge-03** remain unchanged.
```
Comment (jathanism) [inline nautobot/docs/release-notes/version-2.0.md:39]:
```suggestion
        In addition, legacy `Site` instance with name **AMS01** also has three `Device` instances associated with it named **ams01-edge-01**, **ams01-edge-02**,  and **ams01-edge-03**. However, **ams01-edge-01** only has its `site` attribute set as `Site` **AMS01** whereas **ams01-edge-02** and **ams01-edge-03** have both its `site` and `location` attributes set `Site` **AMS01** and `Location` **root-01** respectively. During the data migration, **ams01-edge-01**'s `location` attribute will point to the new `Location` of `LocationType` **Site** with name **AMS01** while devices **ams01-edge-02** and **ams01-edge-03** will remain unchanged.
```

**Review by jathanism — CHANGES_REQUESTED** [review]
I clarified the release notes docs for consistency and readability. Making it so that there is no longer mixed use of `Site` as a model and `Site` as a name, making sure that when each model is referenced it is always `BackTicked`, and if it's a name/value it's **in bold**.

File: nautobot/docs/release-notes/version-2.0.md (hunk 10 above)
Comment (glennmatthews) [reply nautobot/docs/release-notes/version-2.0.md:27]:
```suggestion
3. If you only have `Site` instances in your existing database:
    - A `LocationType` named "Site" will be created and every root level `LocationType` in your database will have the new "Site" `LocationType` as their parent.
    - For each legacy `Site` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created.
    - Model instances that had a `site` field (`CircuitTermination`, `Device`, `PowerPanel`, `RackGroup`, `Rack`, `Prefix`, `VLANGroup`, `VLAN`, `Cluster`) assigned and *did not* have a `location` attribute assigned will be updated to have their `location` point to the new `Location` corresponding to that `Site`. All other attributes on these models will remain unchanged.
```

File: nautobot/docs/release-notes/version-2.0.md (hunk 11 above)
Comment (glennmatthews) [reply nautobot/docs/release-notes/version-2.0.md:30]: Any reason not to use quotes instead of bolding for these name strings?

File: nautobot/docs/release-notes/version-2.0.md (hunk 13 above)
Comment (glennmatthews) [reply nautobot/docs/release-notes/version-2.0.md:33]:
```suggestion
4. If you have both `Site` and `Region` instances in your existing database:
    - A `LocationType` named **Region** will be created.
    - For each legacy `Region` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, etc.) will be created.
    - A `LocationType` named **Site** will be created with the new `LocationType` named **Region** set as its `parent`.
    - Every pre-existing root-level `LocationType` in your database will be updated to have the new `LocationType` named **Site** as its `parent`.
    - For each legacy `Site` instance, a corresponding "Site" `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created with its parent set to the corresponding "Region" `Location` if any.
        - If you have `Site` instances in your database without a `Region` assigned to them, one additional `Location` named **Global Region** of `LocationType` **Region** will be created and each `Location` of `LocationType` **Site** created from the legacy region-less `Site` instances will have the **Global Region** `Location` as their parent.
    - Model instances that had a `site` attribute (`CircuitTermination`, `Device`, `PowerPanel`, `RackGroup`, `Rack`, `Prefix`, `VLANGroup`, `VLAN`, `Cluster`) assigned and *did not* have a `location` attribute assigned will be updated to have their `location` point to the new `Location` of `LocationType` **Site**. All other attributes on these models will remain unchanged.
```

File: nautobot/docs/release-notes/version-2.0.md (hunk 14 above)
Comment (glennmatthews) [reply nautobot/docs/release-notes/version-2.0.md:38]: Consider splitting this into a list of data/actions as well.

File: nautobot/dcim/migrations/0025_migrate_region_and_site_data_to_locations.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+    )
+    for region in regions:
+        location_class.objects.create(
+            location_type=region_lt,
+            name=region.name,
+            description=region.description,
+            parent=location_class.objects.get(name=region.parent.name, location_type=region_lt)
+            if region.parent
+            else None,
+        )
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/dcim/migrations/0025_migrate_region_and_site_data_to_locations.py:29]: The only issue with `bulk_create` is that it does not preserve the top-level-first order in which the locations are created. So this line
```
parent=location_class.objects.get(name=region.parent.name, location_type=region_lt)
            if region.parent
            else None,
```
throws an error when the matching query/parent location does not exist.

We can `bulk_create` all the locations first and then `bulk_update` them with their designated `parent`:
```
location_instances = []
    for region in regions:
        location_instances.append(
            location_class(
                location_type=region_lt,
                name=region.name,
                description=region.description,
            )
        )
location_class.objects.bulk_create(location_instances, 1000)
region_locations = location_class.objects.filter(location_type=region_lt).select_related("parent")
for location in region_locations:
     region = regions.get(name=location.name)
     location.parent = location_class.objects.get(name=region.parent.name) if region.parent else None
location_class.objects.bulk_update(region_locations, ["parent"], 1000)
```
After testing, the second approach is a little slower than the first approach

**Comment by HanlinMiao** [comment]
https://github.com/nautobot/nautobot/pull/3047#discussion_r1064673778. For this comment, it wont let me to reply or quote reply for some reason.

I chose bold letters for no particular reason other than it looks a bit better IMO and it is easier to track than quotes. What do you think?

**Comment by glennmatthews** [comment]
> [#3047 (comment)](https://github.com/nautobot/nautobot/pull/3047#discussion_r1064673778). For this comment, it wont let me to reply or quote reply for some reason.
> 
> I chose bold letters for no particular reason other than it looks a bit better IMO and it is easier to track than quotes. What do you think?

That's fine by me, so long as we're self-consistent.

File: nautobot/docs/release-notes/version-2.0.md [hunk 17]
```diff
@@ -18,6 +18,62 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
 
 #### Collapse Region and Site Models into Location ([#2517](https://github.com/nautobot/nautobot/issues/2517))
 
+##### Initial Data Migration
+
+The `Site` and `Region` models have been removed in v2.0 and have been replaced with `Location` of specific `LocationType`. As a result, the existing `Site` and `Region` data will be migrated to corresponding `LocationType` and `Location` objects. Here is what to expect:
+
+1. If you do not have any `Site` and `Region` instances in your existing database, running this data migration will do nothing.
+2. If you only have `Region` instances in your existing database, a `LocationType` named "Region" will be created and for each legacy `Region` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, etc.) and hierarchy will be created.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.0.md:26]:
```suggestion
2. If you only have `Region` instances in your existing database, a `LocationType` named **Region** will be created and for each legacy `Region` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, etc.) and hierarchy will be created.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 18]
```diff
@@ -18,6 +18,62 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+3. If you only have `Site` instances in your existing database:
+
+    - A `LocationType` named "Site" will be created and every root level `LocationType` in your database will have the new "Site" `LocationType` as their parent.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.0.md:29]:
```suggestion
    - A `LocationType` named **Site** will be created and every preexisting root level `LocationType` in your database will be updated to have the new **Site** `LocationType` as their parent.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 19]
```diff
@@ -18,6 +18,62 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+
+    - For each legacy `Site` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.0.md:31]:
```suggestion
    - For each legacy `Site` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created, and any preexisting `Locations` in your database will be updated to have the appropriate "site" `Locations` as their parents.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 20]
```diff
@@ -18,6 +18,62 @@ Added Site Model Fields to Location. Location Model now has `asn`, `comments`, `
...
+
+    - A `LocationType` named **Region** will be created.
+
+    - For each legacy `Region` instance, a corresponding `Location` instance with the same attributes (`name`, `description`, etc.) will be created.
+
+    - A `LocationType` named **Site** will be created with the new `LocationType` named **Region** set as its `parent`.
+
+    - Every pre-existing root-level `LocationType` in your database will be updated to have the new `LocationType` named **Site** as its `parent`.
+
+    - For each legacy `Site` instance, a corresponding "Site" `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created with its parent set to the corresponding "Region" `Location` if any.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.0.md:53]:
```suggestion
    - For each legacy `Site` instance, a corresponding "site" `Location` instance with the same attributes (`name`, `description`, `tenant`, `facility`, `asn`, `latitude`, `longitude`, etc.) will be created with its parent set to the corresponding "region" `Location` if any.
```

File: nautobot/dcim/migrations/0025_migrate_region_and_site_data_to_locations.py [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+        # If there are 1000 locations loaded into list `location_instances`, we create the locations in batch of 1000
+        # This check is in place to optimize memeroy usage with the creation of large number of location objects
+        if count == 1000:
+            # Handle IntegrityError when Region, Site, Location instances with the same slug.
+            try:
+                location_class.objects.bulk_create(location_instances, batch_size=1000)
+            except IntegrityError as e:
+                logger.error(
+                    f"\n{e.args[0]} \nPlease consider changing the slug attribute of this Region instance to resolve this conflict."
+                )
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/dcim/migrations/0025_migrate_region_and_site_data_to_locations.py:110]: Yeah it seems like even **without** `transaction.atomic`, the migration will not be applied if there is an `IntegrityError` being raised. I tested with some `sites` and `regions` with duplicate slugs and this is the output and confirmed that the `locations` created before the error are rolled back.
Screen Shot 2023-01-10 at 7 25 29 PM

File: nautobot/dcim/migrations/0029_migrate_region_and_site_data_to_locations.py [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+        # If there are 1000 locations loaded into list `location_instances`, we create the locations in batch of 1000
+        # This check is in place to optimize memeroy usage with the creation of large number of location objects
+        if count == 1000:
+            # Handle IntegrityError when Region, Site, Location instances with the same slug.
+            try:
+                location_class.objects.bulk_create(location_instances, batch_size=1000)
+            except IntegrityError as e:
+                logger.error(
+                    f"{e.args[0]} \nPlease consider changing the slug attribute of this Site instance to resolve this conflict."
+                )
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0029_migrate_region_and_site_data_to_locations.py:115]: Should we let the migration fail in this case and similar ones above and below? If we proceed with migrations, seems like there's a data-loss risk.

**Review by glennmatthews — CHANGES_REQUESTED** [review]

File: nautobot/dcim/migrations/0029_migrate_region_and_site_data_to_locations.py (hunk 22 above)
Comment (glennmatthews) [reply nautobot/dcim/migrations/0029_migrate_region_and_site_data_to_locations.py:115]: ah, I misunderstood your earlier comment. So even though we catch the IntegrityError, it still causes the migration to fail? OK then.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/docs/release-notes/version-2.0.md (hunk 11 above)
Comment (jathanism) [reply nautobot/docs/release-notes/version-2.0.md]: To make them stand out.

**Review by jathanism — APPROVED** [review]
:shipit:
